File: backend/services/multi_db_coordinator.py
#!/usr/bin/env python3
"""
═══════════════════════════════════════════════════════════════════════
Agentic Knowledge Integration V2 - Multi-Database Coordinator
═══════════════════════════════════════════════════════════════════════
Date: January 2, 2026
Purpose: Atomic transaction management across 6 databases
═══════════════════════════════════════════════════════════════════════
"""

import os
import uuid
import asyncio
import logging
from datetime import datetime
from typing import Dict, List, Optional, Any, Callable
from dataclasses import dataclass, field
from enum import Enum

# Database clients
import psycopg2
from psycopg2.extras import RealDictCursor
import redis
from qdrant_client import QdrantClient
from neo4j import GraphDatabase
import couchdb

# Local imports
from backend.scripts.setup_redis_v2 import RedisCache

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class TransactionStatus(Enum):
    """Transaction status enum."""
    PENDING = "pending"
    COMMITTED = "committed"
    ROLLED_BACK = "rolled_back"
    FAILED = "failed"


class DatabaseType(Enum):
    """Database type enum."""
    POSTGRESQL = "postgresql"
    NEO4J = "neo4j"
    QDRANT = "qdrant"
    COUCHDB = "couchdb"
    REDIS = "redis"
    MINIO = "minio"


@dataclass
class DBOperation:
    """Database operation with rollback capability."""
    database: DatabaseType
    operation_type: str  # 'insert', 'update', 'delete'
    execute_fn: Callable
    rollback_fn: Callable
    payload: Dict[str, Any]
    executed: bool = False
    result: Optional[Any] = None
    error: Optional[str] = None


@dataclass
class Transaction:
    """Multi-database transaction."""
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    operations: List[DBOperation] = field(default_factory=list)
    status: TransactionStatus = TransactionStatus.PENDING
    created_at: datetime = field(default_factory=datetime.now)
    completed_at: Optional[datetime] = None
    error_message: Optional[str] = None


class MultiDBCoordinator:
    """
    Coordinates atomic transactions across multiple databases.

    Supports: PostgreSQL, Neo4j, Qdrant, CouchDB, Redis, MinIO
    """

    def __init__(
        self,
        postgres_url: Optional[str] = None,
        neo4j_url: Optional[str] = None,
        qdrant_url: Optional[str] = None,
        couchdb_url: Optional[str] = None,
        redis_url: Optional[str] = None,
    ):
        """Initialize database connections."""
        # PostgreSQL
        self.postgres_url = postgres_url or os.getenv(
            "DATABASE_URL", "postgresql://legal_admin:123456@localhost:5434/legal_ai_db"
        )
        self.pg_conn = None

        # Neo4j
        self.neo4j_url = neo4j_url or os.getenv("NEO4J_URL", "bolt://localhost:7687")
        self.neo4j_user = os.getenv("NEO4J_USER", "neo4j")
        self.neo4j_password = os.getenv("NEO4J_PASSWORD", "password")
        self.neo4j_driver = None

        # Qdrant
        self.qdrant_url = qdrant_url or os.getenv("QDRANT_URL", "http://localhost:6333")
        self.qdrant_client = None

        # CouchDB
        self.couchdb_url = couchdb_url or os.getenv(
            "COUCHDB_URL", "http://admin:password@localhost:5984"
        )
        self.couchdb_server = None

        # Redis
        self.redis_url = redis_url or os.getenv("REDIS_URL", "redis://localhost:6379")
        self.redis_cache = None

        # Transaction log
        self.transactions: Dict[str, Transaction] = {}

    def connect(self):
        """Connect to all databases."""
        try:
            # PostgreSQL
            self.pg_conn = psycopg2.connect(self.postgres_url)
            logger.info("✅ PostgreSQL connected")

            # Neo4j
            self.neo4j_driver = GraphDatabase.driver(
                self.neo4j_url,
                auth=(self.neo4j_user, self.neo4j_password)
            )
            logger.info("✅ Neo4j connected")

            # Qdrant
            self.qdrant_client = QdrantClient(url=self.qdrant_url)
            logger.info("✅ Qdrant connected")

            # CouchDB
            self.couchdb_server = couchdb.Server(self.couchdb_url)
            logger.info("✅ CouchDB connected")

            # Redis
            self.redis_cache = RedisCache(url=self.redis_url)
            logger.info("✅ Redis connected")

            logger.info("🎉 All databases connected successfully!")
            return True

        except Exception as e:
            logger.error(f"❌ Database connection failed: {e}")
            raise

    def disconnect(self):
        """Disconnect from all databases."""
        try:
            if self.pg_conn:
                self.pg_conn.close()
            if self.neo4j_driver:
                self.neo4j_driver.close()
            logger.info("✅ All databases disconnected")
        except Exception as e:
            logger.error(f"❌ Disconnect error: {e}")

    def create_transaction(self) -> Transaction:
        """Create a new transaction."""
        transaction = Transaction()
        self.transactions[transaction.id] = transaction
        logger.info(f"📝 Created transaction: {transaction.id}")
        return transaction

    def add_operation(
        self,
        transaction: Transaction,
        database: DatabaseType,
        operation_type: str,
        execute_fn: Callable,
        rollback_fn: Callable,
        payload: Dict[str, Any],
    ):
        """Add an operation to a transaction."""
        operation = DBOperation(
            database=database,
            operation_type=operation_type,
            execute_fn=execute_fn,
            rollback_fn=rollback_fn,
            payload=payload,
        )
        transaction.operations.append(operation)
        logger.info(
            f"➕ Added {database.value} {operation_type} operation to transaction {transaction.id}"
        )

    async def execute_transaction(self, transaction: Transaction) -> bool:
        """
        Execute all operations in a transaction atomically.

        Returns True if all operations succeed, False otherwise.
        Automatically rolls back on failure.
        """
        logger.info(f"🚀 Executing transaction {transaction.id} with {len(transaction.operations)} operations")

        # Log transaction start in PostgreSQL
        await self._log_transaction_start(transaction)

        executed_operations = []

        try:
            # Execute all operations
            for operation in transaction.operations:
                logger.info(
                    f"   ⚙️  Executing {operation.database.value} {operation.operation_type}..."
                )

                try:
                    result = await operation.execute_fn(operation.payload)
                    operation.executed = True
                    operation.result = result
                    executed_operations.append(operation)
                    logger.info(f"   ✅ {operation.database.value} operation succeeded")

                except Exception as e:
                    operation.error = str(e)
                    logger.error(f"   ❌ {operation.database.value} operation failed: {e}")
                    raise

            # All operations succeeded
            transaction.status = TransactionStatus.COMMITTED
            transaction.completed_at = datetime.now()

            # Log transaction commit
            await self._log_transaction_commit(transaction)

            logger.info(f"✅ Transaction {transaction.id} committed successfully!")
            return True

        except Exception as e:
            # Rollback all executed operations
            logger.error(f"❌ Transaction {transaction.id} failed: {e}")
            transaction.status = TransactionStatus.FAILED
            transaction.error_message = str(e)
            transaction.completed_at = datetime.now()

            await self._rollback_operations(executed_operations, transaction)

            # Log transaction rollback
            await self._log_transaction_rollback(transaction)

            return False

    async def _rollback_operations(
        self, operations: List[DBOperation], transaction: Transaction
    ):
        """Rollback executed operations in reverse order."""
        logger.warning(f"🔄 Rolling back {len(operations)} operations...")

        for operation in reversed(operations):
            if operation.executed:
                try:
                    logger.info(f"   ↩️  Rolling back {operation.database.value}...")
                    await operation.rollback_fn(operation.payload, operation.result)
                    logger.info(f"   ✅ {operation.database.value} rollback succeeded")
                except Exception as e:
                    logger.error(f"   ❌ {operation.database.value} rollback failed: {e}")
                    # Continue rolling back other operations

        transaction.status = TransactionStatus.ROLLED_BACK
        logger.info(f"✅ Transaction {transaction.id} rolled back")

    async def _log_transaction_start(self, transaction: Transaction):
        """Log transaction start in PostgreSQL."""
        try:
            cursor = self.pg_conn.cursor()
            cursor.execute(
                """
                INSERT INTO multi_db_transactions (id, operation, status, databases, payload, created_at)
                VALUES (%s, %s, %s, %s, %s, %s)
                """,
                (
                    transaction.id,
                    "multi_db_transaction",
                    transaction.status.value,
                    [op.database.value for op in transaction.operations],
                    None,  # payload as JSONB
                    transaction.created_at,
                ),
            )
            self.pg_conn.commit()
        except Exception as e:
            logger.error(f"Failed to log transaction start: {e}")

    async def _log_transaction_commit(self, transaction: Transaction):
        """Log transaction commit in PostgreSQL."""
        try:
            cursor = self.pg_conn.cursor()
            cursor.execute(
                """
                UPDATE multi_db_transactions
                SET status = %s, completed_at = %s
                WHERE id = %s
                """,
                (transaction.status.value, transaction.completed_at, transaction.id),
            )
            self.pg_conn.commit()
        except Exception as e:
            logger.error(f"Failed to log transaction commit: {e}")

    async def _log_transaction_rollback(self, transaction: Transaction):
        """Log transaction rollback in PostgreSQL."""
        try:
            cursor = self.pg_conn.cursor()
            cursor.execute(
                """
                UPDATE multi_db_transactions
                SET status = %s, completed_at = %s, error_message = %s
                WHERE id = %s
                """,
                (
                    transaction.status.value,
                    transaction.completed_at,
                    transaction.error_message,
                    transaction.id,
                ),
            )
            self.pg_conn.commit()
        except Exception as e:
            logger.error(f"Failed to log transaction rollback: {e}")

    def get_transaction_stats(self) -> Dict[str, Any]:
        """Get transaction statistics."""
        total = len(self.transactions)
        committed = sum(1 for t in self.transactions.values() if t.status == TransactionStatus.COMMITTED)
        rolled_back = sum(1 for t in self.transactions.values() if t.status == TransactionStatus.ROLLED_BACK)
        failed = sum(1 for t in self.transactions.values() if t.status == TransactionStatus.FAILED)
        pending = sum(1 for t in self.transactions.values() if t.status == TransactionStatus.PENDING)

        return {
            "total_transactions": total,
            "committed": committed,
            "rolled_back": rolled_back,
            "failed": failed,
            "pending": pending,
            "success_rate": round(committed / total * 100, 2) if total > 0 else 0,
        }


# Example usage
async def example_usage():
    """Example of using the MultiDBCoordinator."""
    coordinator = MultiDBCoordinator()
    coordinator.connect()

    # Create a transaction
    transaction = coordinator.create_transaction()

    # Define operations
    async def insert_postgres(payload):
        cursor = coordinator.pg_conn.cursor()
        cursor.execute(
            "INSERT INTO enhanced_tags (id, name, category, file_path, timestamp) VALUES (%s, %s, %s, %s, %s) RETURNING id",
            (payload["id"], payload["name"], payload["category"], payload["file_path"], datetime.now()),
        )
        coordinator.pg_conn.commit()
        return cursor.fetchone()[0]

    async def rollback_postgres(payload, result):
        cursor = coordinator.pg_conn.cursor()
        cursor.execute("DELETE FROM enhanced_tags WHERE id = %s", (result,))
        coordinator.pg_conn.commit()

    async def insert_qdrant(payload):
        from qdrant_client.models import PointStruct
        coordinator.qdrant_client.upsert(
            collection_name="knowledge_base_v2",
            points=[
                PointStruct(
                    id=payload["id"],
                    vector=payload["embedding"],
                    payload={"name": payload["name"], "category": payload["category"]},
                )
            ],
        )
        return payload["id"]

    async def rollback_qdrant(payload, result):
        coordinator.qdrant_client.delete(
            collection_name="knowledge_base_v2",
            points_selector=[result],
        )

    # Add operations to transaction
    test_payload = {
        "id": str(uuid.uuid4()),
        "name": "test_tag",
        "category": "file",
        "file_path": "/test/file.ts",
        "embedding": [0.1] * 384,
    }

    coordinator.add_operation(
        transaction,
        DatabaseType.POSTGRESQL,
        "insert",
        insert_postgres,
        rollback_postgres,
        test_payload,
    )

    coordinator.add_operation(
        transaction,
        DatabaseType.QDRANT,
        "insert",
        insert_qdrant,
        rollback_qdrant,
        test_payload,
    )

    # Execute transaction
    success = await coordinator.execute_transaction(transaction)

    if success:
        print("✅ Transaction completed successfully!")
    else:
        print("❌ Transaction failed and was rolled back")

    # Get stats
    stats = coordinator.get_transaction_stats()
    print(f"\n📊 Transaction Stats: {stats}")

    coordinator.disconnect()


if __name__ == "__main__":
    asyncio.run(example_usage())

File: deeds_labs/python-middleware/backend/services/multi_db_coordinator.ts
/**
 * ═══════════════════════════════════════════════════════════════════════
 * Agentic Knowledge Integration V2 - Multi-Database Coordinator (TypeScript)
 * ═══════════════════════════════════════════════════════════════════════
 * Date: January 2, 2026
 * Purpose: Atomic transaction management across 6 databases
 * ═══════════════════════════════════════════════════════════════════════
 */

import { Pool } from 'pg';
import { createClient as createRedisClient } from 'redis';
import { QdrantClient } from '@qdrant/js-client-rest';
import neo4j from 'neo4j-driver';
import { v4 as uuidv4 } from 'uuid';

export enum TransactionStatus {
  PENDING = 'pending',
  COMMITTED = 'committed',
  ROLLED_BACK = 'rolled_back',
  FAILED = 'failed',
}

export enum DatabaseType {
  POSTGRESQL = 'postgresql',
  NEO4J = 'neo4j',
  QDRANT = 'qdrant',
  COUCHDB = 'couchdb',
  REDIS = 'redis',
  MINIO = 'minio',
}

export interface DBOperation<T = any, R = any> {
  database: DatabaseType;
  operationType: 'insert' | 'update' | 'delete';
  executeFn: (payload: T) => Promise<R>;
  rollbackFn: (payload: T, result?: R) => Promise<void>;
  payload: T;
  executed?: boolean;
  result?: R;
  error?: string;
}

export interface Transaction {
  id: string;
  operations: DBOperation[];
  status: TransactionStatus;
  createdAt: Date;
  completedAt?: Date;
  errorMessage?: string;
}

export interface TransactionStats {
  totalTransactions: number;
  committed: number;
  rolledBack: number;
  failed: number;
  pending: number;
  successRate: number;
}

export interface MultiDBCoordinatorConfig {
  postgresUrl?: string;
  neo4jUrl?: string;
  neo4jUser?: string;
  neo4jPassword?: string;
  qdrantUrl?: string;
  couchdbUrl?: string;
  redisUrl?: string;
}

export class MultiDBCoordinator {
  private pgPool: Pool;
  private neo4jDriver: neo4j.Driver;
  private qdrantClient: QdrantClient;
  private redisClient: ReturnType<typeof createRedisClient>;
  private transactions: Map<string, Transaction> = new Map();
  private connected: boolean = false;

  constructor(config: MultiDBCoordinatorConfig = {}) {
    // PostgreSQL
    this.pgPool = new Pool({
      connectionString:
        config.postgresUrl ||
        process.env.DATABASE_URL ||
        'postgresql://legal_admin:123456@localhost:5434/legal_ai_db',
    });

    // Neo4j
    const neo4jUrl = config.neo4jUrl || process.env.NEO4J_URL || 'bolt://localhost:7687';
    const neo4jUser = config.neo4jUser || process.env.NEO4J_USER || 'neo4j';
    const neo4jPassword = config.neo4jPassword || process.env.NEO4J_PASSWORD || 'password';
    this.neo4jDriver = neo4j.driver(neo4jUrl, neo4j.auth.basic(neo4jUser, neo4jPassword));

    // Qdrant
    const qdrantUrl = config.qdrantUrl || process.env.QDRANT_URL || 'http://localhost:6333';
    this.qdrantClient = new QdrantClient({ url: qdrantUrl });

    // Redis
    const redisUrl = config.redisUrl || process.env.REDIS_URL || 'redis://localhost:6379';
    this.redisClient = createRedisClient({ url: redisUrl });

    this.redisClient.on('error', (err) => {
      console.error('Redis Client Error:', err);
    });
  }

  async connect(): Promise<void> {
    try {
      // Test PostgreSQL
      await this.pgPool.query('SELECT 1');
      console.log('✅ PostgreSQL connected');

      // Test Neo4j
      await this.neo4jDriver.verifyConnectivity();
      console.log('✅ Neo4j connected');

      // Test Qdrant
      await this.qdrantClient.getCollections();
      console.log('✅ Qdrant connected');

      // Connect Redis
      await this.redisClient.connect();
      console.log('✅ Redis connected');

      this.connected = true;
      console.log('🎉 All databases connected successfully!');
    } catch (error) {
      console.error('❌ Database connection failed:', error);
      throw error;
    }
  }

  async disconnect(): Promise<void> {
    try {
      await this.pgPool.end();
      await this.neo4jDriver.close();
      await this.redisClient.quit();
      this.connected = false;
      console.log('✅ All databases disconnected');
    } catch (error) {
      console.error('❌ Disconnect error:', error);
    }
  }

  createTransaction(): Transaction {
    const transaction: Transaction = {
      id: uuidv4(),
      operations: [],
      status: TransactionStatus.PENDING,
      createdAt: new Date(),
    };

    this.transactions.set(transaction.id, transaction);
    console.log(`📝 Created transaction: ${transaction.id}`);
    return transaction;
  }

  addOperation<T = any, R = any>(
    transaction: Transaction,
    database: DatabaseType,
    operationType: 'insert' | 'update' | 'delete',
    executeFn: (payload: T) => Promise<R>,
    rollbackFn: (payload: T, result?: R) => Promise<void>,
    payload: T
  ): void {
    const operation: DBOperation<T, R> = {
      database,
      operationType,
      executeFn,
      rollbackFn,
      payload,
      executed: false,
    };

    transaction.operations.push(operation);
    console.log(
      `➕ Added ${database} ${operationType} operation to transaction ${transaction.id}`
    );
  }

  async executeTransaction(transaction: Transaction): Promise<boolean> {
    console.log(
      `🚀 Executing transaction ${transaction.id} with ${transaction.operations.length} operations`
    );

    // Log transaction start
    await this.logTransactionStart(transaction);

    const executedOperations: DBOperation[] = [];

    try {
      // Execute all operations
      for (const operation of transaction.operations) {
        console.log(`   ⚙️  Executing ${operation.database} ${operation.operationType}...`);

        try {
          const result = await operation.executeFn(operation.payload);
          operation.executed = true;
          operation.result = result;
          executedOperations.push(operation);
          console.log(`   ✅ ${operation.database} operation succeeded`);
        } catch (error) {
          operation.error = error instanceof Error ? error.message : String(error);
          console.error(`   ❌ ${operation.database} operation failed:`, error);
          throw error;
        }
      }

      // All operations succeeded
      transaction.status = TransactionStatus.COMMITTED;
      transaction.completedAt = new Date();

      // Log transaction commit
      await this.logTransactionCommit(transaction);

      console.log(`✅ Transaction ${transaction.id} committed successfully!`);
      return true;
    } catch (error) {
      // Rollback all executed operations
      console.error(`❌ Transaction ${transaction.id} failed:`, error);
      transaction.status = TransactionStatus.FAILED;
      transaction.errorMessage = error instanceof Error ? error.message : String(error);
      transaction.completedAt = new Date();

      await this.rollbackOperations(executedOperations, transaction);

      // Log transaction rollback
      await this.logTransactionRollback(transaction);

      return false;
    }
  }

  private async rollbackOperations(
    operations: DBOperation[],
    transaction: Transaction
  ): Promise<void> {
    console.warn(`🔄 Rolling back ${operations.length} operations...`);

    // Rollback in reverse order
    for (const operation of operations.reverse()) {
      if (operation.executed) {
        try {
          console.log(`   ↩️  Rolling back ${operation.database}...`);
          await operation.rollbackFn(operation.payload, operation.result);
          console.log(`   ✅ ${operation.database} rollback succeeded`);
        } catch (error) {
          console.error(`   ❌ ${operation.database} rollback failed:`, error);
          // Continue rolling back other operations
        }
      }
    }

    transaction.status = TransactionStatus.ROLLED_BACK;
    console.log(`✅ Transaction ${transaction.id} rolled back`);
  }

  private async logTransactionStart(transaction: Transaction): Promise<void> {
    try {
      await this.pgPool.query(
        `
        INSERT INTO multi_db_transactions (id, operation, status, databases, created_at)
        VALUES ($1, $2, $3, $4, $5)
        `,
        [
          transaction.id,
          'multi_db_transaction',
          transaction.status,
          transaction.operations.map((op) => op.database),
          transaction.createdAt,
        ]
      );
    } catch (error) {
      console.error('Failed to log transaction start:', error);
    }
  }

  private async logTransactionCommit(transaction: Transaction): Promise<void> {
    try {
      await this.pgPool.query(
        `
        UPDATE multi_db_transactions
        SET status = $1, completed_at = $2
        WHERE id = $3
        `,
        [transaction.status, transaction.completedAt, transaction.id]
      );
    } catch (error) {
      console.error('Failed to log transaction commit:', error);
    }
  }

  private async logTransactionRollback(transaction: Transaction): Promise<void> {
    try {
      await this.pgPool.query(
        `
        UPDATE multi_db_transactions
        SET status = $1, completed_at = $2, error_message = $3
        WHERE id = $4
        `,
        [transaction.status, transaction.completedAt, transaction.errorMessage, transaction.id]
      );
    } catch (error) {
      console.error('Failed to log transaction rollback:', error);
    }
  }

  getTransactionStats(): TransactionStats {
    const transactions = Array.from(this.transactions.values());
    const total = transactions.length;
    const committed = transactions.filter((t) => t.status === TransactionStatus.COMMITTED).length;
    const rolledBack = transactions.filter(
      (t) => t.status === TransactionStatus.ROLLED_BACK
    ).length;
    const failed = transactions.filter((t) => t.status === TransactionStatus.FAILED).length;
    const pending = transactions.filter((t) => t.status === TransactionStatus.PENDING).length;

    return {
      totalTransactions: total,
      committed,
      rolledBack,
      failed,
      pending,
      successRate: total > 0 ? Math.round((committed / total) * 10000) / 100 : 0,
    };
  }

  getTransaction(id: string): Transaction | undefined {
    return this.transactions.get(id);
  }

  // Convenience methods for common operations

  async insertEnhancedTag(tagData: {
    id: string;
    name: string;
    category: string;
    filePath: string;
    embedding: number[];
    summary?: string;
  }): Promise<boolean> {
    const transaction = this.createTransaction();

    // PostgreSQL operation
    this.addOperation(
      transaction,
      DatabaseType.POSTGRESQL,
      'insert',
      async (payload) => {
        const result = await this.pgPool.query(
          `
          INSERT INTO enhanced_tags (id, name, category, file_path, summary, timestamp)
          VALUES ($1, $2, $3, $4, $5, $6)
          RETURNING id
          `,
          [payload.id, payload.name, payload.category, payload.filePath, payload.summary, new Date()]
        );
        return result.rows[0].id;
      },
      async (payload, result) => {
        await this.pgPool.query('DELETE FROM enhanced_tags WHERE id = $1', [result]);
      },
      tagData
    );

    // Qdrant operation
    this.addOperation(
      transaction,
      DatabaseType.QDRANT,
      'insert',
      async (payload) => {
        await this.qdrantClient.upsert('knowledge_base_v2', {
          points: [
            {
              id: payload.id,
              vector: payload.embedding,
              payload: {
                tag_id: payload.id,
                name: payload.name,
                category: payload.category,
                file_path: payload.filePath,
                summary: payload.summary,
                timestamp: new Date().toISOString(),
              },
            },
          ],
        });
        return payload.id;
      },
      async (payload, result) => {
        await this.qdrantClient.delete('knowledge_base_v2', {
          points: [result],
        });
      },
      tagData
    );

    return await this.executeTransaction(transaction);
  }
}

// Singleton instance
let coordinatorInstance: MultiDBCoordinator | null = null;

export function getMultiDBCoordinator(config?: MultiDBCoordinatorConfig): MultiDBCoordinator {
  if (!coordinatorInstance) {
    coordinatorInstance = new MultiDBCoordinator(config);
  }
  return coordinatorInstance;
}

// Example usage
export async function exampleUsage() {
  const coordinator = getMultiDBCoordinator();
  await coordinator.connect();

  // Insert an enhanced tag across PostgreSQL and Qdrant
  const success = await coordinator.insertEnhancedTag({
    id: uuidv4(),
    name: 'TestComponent',
    category: 'component',
    filePath: '/src/components/TestComponent.svelte',
    embedding: new Array(384).fill(0.1),
    summary: 'A test component for demonstration',
  });

  console.log('Transaction result:', success ? '✅ Success' : '❌ Failed');

  // Get statistics
  const stats = coordinator.getTransactionStats();
  console.log('Transaction stats:', stats);

  await coordinator.disconnect();
}
